src/std/net/s3/api.ss

```diff
@@ -0,0 +1,242 @@
+           ; buckets is #f if none are returned
+           (request-close req)
+           names))))
+
+;; NOTE: all bucket operations need the correct region for the bucket or they will 400
+(defmethod {create-bucket! s3-client}
+  (lambda (self bucket)
+    (using (self :- s3-client)
+      (let (req (s3-request/error self verb: 'PUT bucket: bucket))
+        (request-close req)
+        (void)))))
+
+; Gets a bucket struct that can be used to fetch objects.
+(defmethod {get-bucket s3-client} ; => bucket
+  (lambda (self bucket-name)
+    (using (self :- s3-client)
+      (if (s3-client::bucket-exists? self bucket-name)
+        (make-bucket self bucket-name self.region)
+        #f))))
+
+; Delete a bucket by name
+(defmethod {delete-bucket! s3-client}
+  (lambda (self bucket)
+    (using (self :- s3-client)
+      (when (s3-client::bucket-exists? self bucket)
+        (let (req (s3-request/error self verb: 'DELETE bucket: bucket))
+          (request-close req)
+          (void))))))
+
+(defmethod {bucket-exists? s3-client}
+  (lambda (self bucket)
+    (using (self :- s3-client)
+           (let* ((bucket (if (bucket? bucket) (bucket-name bucket) bucket))
+                  (req {self.request verb: 'HEAD bucket: bucket})
+                  (code (request-status req)))
+             ; 200 and 404 are expected codes
+             ; we explicitly handle 404 so we get proper predicate
+             ; semantics and don't raise on what would otherwise be
+             ; #f condition.
+             (if (memv code [200 404]) 
+               (begin
+                 (request-close req)
+                 (= code 200))
+               (with-request-error req))))))
+
+(defmethod {bucket s3-client}
+  (lambda (self name)
+    (using (self self :- s3-client)
+      (if (s3-client::bucket-exists? self name)
+        (make-bucket self name (s3-client-region self))
+        #f))))
+
+; Lists the objects stored within the bucket
+(defmethod {list-objects bucket}
+  (lambda (self)
+    (using ((self self :- bucket)
+            (client (bucket-client self) : s3-client))
+           (let* ((name (bucket-name self))
+                  (req (s3-request/error client verb: 'GET bucket: name))
+                  (xml (s3-parse-xml req))
+                  (keys (sxml-select xml (sxml-e? 's3:Key) cadr)))
+             (request-close req)
+             keys))))
+
+(defmethod {get bucket}
+  (lambda (self key)
+    (using ((self self : bucket)
+            (client (bucket-client self) : s3-client))
+      (let* ((req (s3-request/error client verb: 'GET bucket: (bucket-name self)
+                                    path: (string-append "/" key)))
+             (data (request-content req)))
+        (request-close req)
+        data))))
+
+(defmethod {put! bucket}
+  (lambda (self key data content-type: (content-type "binary/octet-stream"))
+    (using ((self self : bucket)
+            (client (bucket-client self) : s3-client))
```

Comment:
self is known to be a bucket here, so `:-`.
And client is known to be an `s3-client`, so `:-` again.

Reply:
Also, you can write `self.client` from the first using, remember it has `let*` semantics.